lib/SILOptimizer/Mandatory/AddressLowering.h

```diff
@@ -0,0 +1,289 @@
+/// 3. Materialized. 'storageAddress' is valid. Address projections have been
+/// emitted at the point that this value is defined.
+///
+/// 4. Rewritten. The definition of this opaque value is fully translated
+/// into lowered SIL. Instructions are typically materialized and rewritten at
+/// the same time. A indirect result, however, is materialized as soon as its
+/// alloc_stack is emitted, but only rewritten once the call itself is
+/// rewritten.
+///
+/// A projection may project out of an operand's definition (def-projection).
+/// After allocation, before materialization or rewriting, we may have:
+///
+///   %result_addr = alloc_stack       // storage for %result
+///   %result = apply : $() -> @out T
+///   %extract = struct_extact %result // def-projection of %result
+///
+/// Or, a projection may project into a composing use (use-projection):
+///
+///   %struct_addr = alloc_stack      // storage for %struct
+///   %result = apply : $() -> @out T // use-projection of %struct at operand #0
+///   %struct = struct (%result)
+///
+/// A phi-projection is a use projection that projects its entire value
+/// through a phi rather than into a composing use. It has an invalid
+/// 'projectedOperandNum':
+///
+///     %result = apply : $() -> @out T // use-projection of %phi
+///     br bb1(%result)
+///   bb1(%phi : @owned $T)
+///
+/// Operations that destructively reuse storage (open_existential_value,
+/// unchecked_enum_data, and switch_enum), are not considered storage
+/// projections. Instead, these values have no ValueStorage but are rewritten to
+/// directly reuse their operand's storage.
+///
+/// To materialize projections, address lowering follows the original def-use
+/// edges for opaque values. Consequently, values that have storage cannot
+/// be removed from SIL or from the storage map until rewriting is
+/// complete. Mapped values can, however, be substituted on-the-fly by emitting
+/// a place-holder value and updating the map entry. This works because the
+/// value storage map holds no direct references to any SIL entities, such as
+/// Operands or SILValues.
+struct ValueStorage {
+  enum : uint32_t { InvalidID = uint32_t(~0) };
+  enum : uint16_t { InvalidOper = uint16_t(~0) };
+
+  /// The final address of this storage after rewriting the SIL. For values
+  /// linked to their own storage, this is set during storage allocation to an
+  /// alloc_stack or indirect function argument. For projections, it is only set
+  /// after materialization (during instruction rewriting).
+  SILValue storageAddress;
+
+  /// When either isDefProjection or isUseProjection is set, this refers to the
+  /// storage whose "def" this value projects out of or whose operand this
+  /// storage projects into via its "use.
+  uint32_t projectedStorageID = InvalidID;
+
+  /// For use-projections, identifies the operand index of the composing use.
+  /// Only valid for non-phi use projections.
+  uint16_t projectedOperandNum = InvalidOper;
```

Comment:
> There are (crazy) situations where 16 bits are not enough. E.g. see https://github.com/apple/swift/pull/35289

@eeckstein In this case, I'll simply avoid projecting storage into a use beyond 64k operands. It is worthwhile for ValueStorage to be 2 words, and introducing a PointerIntPair is not worthwhile. Note that this actually still optimizes the preposterous test case you pointed out, because that type is nested.

In general, the way to handle the problem of importing a static array larger than 64k is definitely not to use 32-bits for an operand index. This will blow up the compiler in any many places where the complexity is cubic or worse when operands are assumed to scale proportionally with the function size.

The only hope of making the compiler handle types this large is to *actively* guard against analyzing operands lists larger then 64K. That means going through all the -Onone passes and inserting checks to bypass any analysis of those instructions.